src/lua/protobuf.lua

```diff
@@ -0,0 +1,662 @@
+        for _, exp_t in pairs(exp_type) do
+            if type(value) == exp_t then
+                found = true
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_range(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    -- If one of the limits is 'nil' this function skips
+    -- the comparison with this limit
+    --
+    -- NB: We can't use -math.huge instead of nil for the lower limit,
+    -- because, for example, 10ULL < -math.huge returns true.
+    if min ~= nil and value < min or max ~= nil and value > max then
+        error(('Input data for %q field is %q and do not fit in %q')
+            :format(field_def.name, value, field_def.type))
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    local value_type = type(value)
+    assert(scalar_def.accept_type ~= nil)
+    -- Checks type of input according to the allowed types for this field.
+    validate_type(field_def, value, scalar_def.accept_type)
+    -- Checks length of the string if input type assumed to be string.
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    -- Checks number values for being NaN or inf.
+    if value_type == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if value_type == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    -- Checks values assumed to be integer.
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    -- Checks numeric values to see if they belong to the limits.
+    if scalar_def.limits ~= nil then
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_range(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Encoders
+
+local function encode_repeated(protocol, field_def, value)
+    local buf = {}
+    local encode_as_packed = false
+    if type(value) ~= 'table' then
+        error('For repeated fields table data are needed')
+    end
+    validate_table_is_array(field_def, value)
+    if is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        encode_as_packed = scalar_def.encode_as_packed
+    end
+    for _, item in ipairs(value) do
```

Comment:
The same question: if an item is `box.NULL`, what is the behavior? Should the `validate_table_is_array` check takes such values into account?

Reply:
As I checked `box.NULL` is not equal to a missing field in array. So I suppose that current implementation will parse `box.NULL` in `repeated` field. 

```
tarantool> for key, value in pairs(a) do
print(key, value)
end
1       1
3       3
---
...

tarantool> a = {1, box.NULL, 3}
---
...

tarantool> for key, value in pairs(a) do
print(key, value)
end
1       1
2       cdata<void *>: NULL
3       3
---
...
```
But when viewed from the decoder side if we get `{1, 3}` we can't determine which key field `3` has, so I suppose we have to `error()` this.

Reply:
I agree, let's give an error. At least we can reconsider this behavior in a future.